pkg/kubelet/kubeletconfig/checkpoint/store/fsstore.go

```diff
@@ -21,82 +21,99 @@ import (
-	checkpointsDir string
+	// dir is the absolute path to the storage directory for fsStore
+	dir string
 }
 
-// NewFsStore returns a Store that saves its data in `checkpointsDir`
-func NewFsStore(fs utilfs.Filesystem, checkpointsDir string) Store {
+var _ Store = (*fsStore)(nil)
+
+// NewFsStore returns a Store that saves its data in dir
+func NewFsStore(fs utilfs.Filesystem, dir string) Store {
 	return &fsStore{
-		fs:             fs,
-		checkpointsDir: checkpointsDir,
+		fs:  fs,
+		dir: dir,
 	}
 }
 
 func (s *fsStore) Initialize() error {
-	utillog.Infof("initializing config checkpoints directory %q", s.checkpointsDir)
-	if err := utilfiles.EnsureDir(s.fs, s.checkpointsDir); err != nil {
+	utillog.Infof("initializing config checkpoints directory %q", s.dir)
+	// ensure top-level dir for store
+	if err := utilfiles.EnsureDir(s.fs, s.dir); err != nil {
+		return err
+	}
+	// ensure metadata directory and reference files (tracks current and lkg configs)
+	if err := utilfiles.EnsureDir(s.fs, filepath.Join(s.dir, metaDir)); err != nil {
+		return err
+	}
+	if err := utilfiles.EnsureFile(s.fs, s.metaPath(currentFile)); err != nil {
 		return err
 	}
-	if err := utilfiles.EnsureFile(s.fs, filepath.Join(s.checkpointsDir, curFile)); err != nil {
+	if err := utilfiles.EnsureFile(s.fs, s.metaPath(lastKnownGoodFile)); err != nil {
 		return err
 	}
-	return utilfiles.EnsureFile(s.fs, filepath.Join(s.checkpointsDir, lkgFile))
+	// ensure checkpoints directory (saves unpacked payloads in subdirectories named after payload UID)
+	return utilfiles.EnsureDir(s.fs, filepath.Join(s.dir, checkpointsDir))
 }
 
-func (s *fsStore) Exists(uid string) (bool, error) {
-	ok, err := utilfiles.FileExists(s.fs, filepath.Join(s.checkpointsDir, uid))
+func (s *fsStore) Exists(c checkpoint.RemoteConfigSource) (bool, error) {
+	// we check whether the directory was created for the resource
+	uid := c.UID()
+	ok, err := utilfiles.DirExists(s.fs, s.checkpointPath(uid))
 	if err != nil {
 		return false, fmt.Errorf("failed to determine whether checkpoint %q exists, error: %v", uid, err)
 	}
 	return ok, nil
 }
 
-func (s *fsStore) Save(c checkpoint.Checkpoint) error {
-	// encode the checkpoint
-	data, err := c.Encode()
-	if err != nil {
-		return err
-	}
-	// save the file
-	return utilfiles.ReplaceFile(s.fs, filepath.Join(s.checkpointsDir, c.UID()), data)
+func (s *fsStore) Save(c checkpoint.Payload) error {
+	// save the checkpoint's files in the appropriate checkpoint dir
+	return utilfiles.ReplaceDir(s.fs, s.checkpointPath(c.UID()), c.Files())
 }
 
-func (s *fsStore) Load(uid string) (checkpoint.Checkpoint, error) {
-	filePath := filepath.Join(s.checkpointsDir, uid)
-	utillog.Infof("loading configuration from %q", filePath)
-
-	// load the file
-	data, err := s.fs.ReadFile(filePath)
+func (s *fsStore) Load(source checkpoint.RemoteConfigSource) (*kubeletconfig.KubeletConfiguration, error) {
+	sourceFmt := fmt.Sprintf("%s:%s", source.APIPath(), source.UID())
```

Comment:
is `:` a safe filesystem character? on windows as well?

Reply:
Ah, darn, not safe on Windows. I'll update. Thanks.

Reply:
Oh, hold on, this is just for logging. Not written to fs. Too long since I last read this code xD.

Reply:
ah, misread